### PR TITLE
routing/rpl: adjusted RPL to use the FIB

### PR DESCRIPTION
**Pull request by BytesGalore**
Formerly the RPL implementation used internal routing and forwarding tables.
This PR changes it to use the new FIB instead of providing a `get_next_hop_provider()` to the current fading sixlowpan/ip implementation.
The provider registration and usage is replaced by calling the FIB for a next-hop directly.

### CODE REVIEW

File: sys/net/include/rpl.h [hunk 1]
```diff
@@ -261,6 +204,13 @@ rpl_routing_entry_t *rpl_get_routing_table(void);
  * */
 uint8_t rpl_is_root(void);
 
+/**
+ * @brief receive the set inteface id
+ *
+ * @return the set interface id of KERNEL_PID_UNDEF if no interface has been set
```
Comment (cgundogan) [inline sys/net/include/rpl.h]: s/of/or

File: sys/net/routing/rpl/rpl.c [hunk 2]
```diff
@@ -116,6 +110,11 @@ uint8_t rpl_init(int if_id, ipv6_addr_t *address)
     return SIXLOWERROR_SUCCESS;
 }
 
+kernel_pid_t rpl_get_if(void)
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl.c]: I remember a conversation with  @OlegHahm citing @Kijewski that C is not Java (: Maybe expose rpl_if_id instead of using getter/setter?

File: sys/net/routing/rpl/rpl_control_messages.c [hunk 3]
```diff
@@ -840,8 +840,17 @@ void rpl_recv_DAO(void)
                       ipv6_addr_to_str(addr_str, IPV6_MAX_ADDR_STR_LEN, &rpl_opt_target_buf->target),
                       ipv6_addr_to_str(addr_str, IPV6_MAX_ADDR_STR_LEN, &ipv6_buf->srcaddr),
                       (rpl_opt_transit_buf->path_lifetime * my_dodag->lifetime_unit));
-                rpl_add_routing_entry(&rpl_opt_target_buf->target, &ipv6_buf->srcaddr,
-                                      rpl_opt_transit_buf->path_lifetime * my_dodag->lifetime_unit);
+
+                /* If we are the target we don't need to add a new FIB entry
+                 * This is a bit hack-ish, but sufficient for now
+                */
+                if ( my_address.uint32[3] != rpl_opt_target_buf->target.uint32[3] ) {
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: why only check the last 4 bytes and not the full address? You could use
```C
static inline int ipv6_addr_is_equal(const ipv6_addr_t *a, const ipv6_addr_t *b)
```
from ip.h

File: sys/net/routing/rpl/rpl_dodag.c [hunk 4]
```diff
@@ -164,7 +165,11 @@ bool rpl_equal_id(ipv6_addr_t *id1, ipv6_addr_t *id2)
     }
 
     return true;
+}
 
+rpl_parent_t *rpl_get_parents(void)
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl_dodag.c]: the same discussion as `rpl_get_if_id()`. Expose maybe?

File: sys/shell/commands/sc_fib.c [hunk 5]
```diff
@@ -71,7 +71,7 @@ int _fib_route_handler(int argc, char **argv)
     /* e.g. fibroute right now dont care about the adress/protocol family */
     if (argc == 1) {
         fib_print_routes();
-        return;
+        return 0;
```
Comment (cgundogan) [inline sys/shell/commands/sc_fib.c]: I think changes to this file are independent from your RPL refactorings and could be outsourced to its own PR. Reducing the diff slightly.

File: sys/net/routing/rpl/rpl.c (hunk 2 above)
Comment (authmillenon) [inline sys/net/routing/rpl/rpl.c]: Moreover: which pid is returned here? The interface representing RPL? Or the interface RPL is registered to (as in "is a router for"). The first case doesn't make sense, since interfaces are protocol threads that IP sends stuff over. The second case disregards the possibility to have RPL on more than one interface. Solution could be to give RPL its own interface representation (similar to `ng_ipv6_netif`).

**Comment by cgundogan** [comment]
+137 −294. This is nice (: Need to do more testing and reviewing, though.

File: sys/net/network_layer/sixlowpan/ip.c [hunk 6]
```diff
@@ -33,6 +33,7 @@
 #include "lowpan.h"
```
Comment (cgundogan) [inline sys/net/network_layer/sixlowpan/ip.c]: Does it also make sense to outsource the changes to this diff to its own PR? Doesn't seem like that there are any dependencies to your RPL refactorings?

File: sys/net/routing/rpl/rpl.c (hunk 2 above)
Comment (cgundogan) [inline sys/net/routing/rpl/rpl.c]: @authmillenon Since this is still the old ip/sixlo stack there is only this *global independent thread approach* without any ties to interfaces. Am I wrong?

Comment (authmillenon) [inline sys/net/routing/rpl/rpl.c]: Ah okay… I was confused because the old stack uses just plain integers as interfaces (which are enumerated 0…N; because there are no extra threads for interfaces, just the transceiver thread) while the new stack just uses the PID of the MAC layer.

File: sys/net/network_layer/sixlowpan/ip.c (hunk 6 above)
Comment (BytesGalore) [inline sys/net/network_layer/sixlowpan/ip.c]: Basically yes, but then no routing protocol will work in mainline RIOT until we merge this PR or adjust the other protocols.

File: sys/net/routing/rpl/rpl_control_messages.c (hunk 3 above)
Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: @cgundogan the reason for this hack-ish check is that `my_addres` is the global IPv6 address, e.g. `abcd::ff:fe00:1234` and the `rpl_opt_target_buf->target` the link local address, e.g. `fe80::ff:fe00:1234`.
I'm also in favour using the proposed function, but I didn't find a way to ask for the link local address of the node here ( I also didn't searched for long ;) )

File: sys/shell/commands/sc_fib.c (hunk 5 above)
Comment (BytesGalore) [inline sys/shell/commands/sc_fib.c]: agree, I opened #2782 and will remove the changes from here.

**Comment by BytesGalore** [comment]
does not depend, but would make sense to merge after #2782 has been merged, since [this](https://github.com/BytesGalore/RIOT/commit/b555f6057f5548691e043afb4f5e6e3050b9fd3b) "compatibility" commit can be removed then.

File: sys/net/routing/rpl/rpl.c [hunk 7]
```diff
@@ -279,32 +273,16 @@ void *rpl_process(void *arg)
 void _rpl_update_routing_table(void)
 {
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl.c]: Since FIB handles the lifetime of routes, would it make sense to remove this `_rpl_update_routing_table` thread completely and just pass a function callback to the fib, which should be executed when a certain condition holds? (e.g. route lifetime expired) @BytesGalore @OlegHahm

File: sys/net/routing/rpl/rpl_control_messages.c (hunk 3 above)
Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: Wireshark shows that the `target` field in the `RPL TARGET Option` also is a global address. There was a PR some weeks ago introducing this behavior

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: Hmm, when I enable `DEBUG`, `rpl_opt_target_buf->target` reveals the link local address ~5 lines above.
Maybe there's something wrong with the `rpl_opt_target_buf` construction/access

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: I dumped the `rpl_opt_target_buf` of a packet and it shows that the stored address is the link local one. This applies to `native` and `iot-lab_M3`

File: sys/net/routing/rpl/rpl.c (hunk 7 above)
Comment (BytesGalore) [inline sys/net/routing/rpl/rpl.c]: @cgundogan the FIB works completely without threads. The exact invalidation point in time of a route is fixed and known, but not monitored by the FIB. A route lifetime and its invalidation is only performed on demand when the specific FIB entry is touched.
So if the lifetime of a route exceeds, probably the parent invalidation will not be triggered.

File: sys/net/routing/rpl/rpl_control_messages.c (hunk 3 above)
Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: you are right, I dumped the target using your branch. On master however, a global ipv6 address is displayed. Could you rebase, please?

**Comment by cgundogan** [comment]
I had some warnings compiling your branch regarding `return` values in `sc_fib.c`. But I realized that this is fixed in master. A rebase would be appreciated (:

**Comment by BytesGalore** [comment]
rebased

**Comment by cgundogan** [comment]
@BytesGalore did you rebase to current master? I still get the warning regarding `sc_fib.c` and `rpl_send_opt_target_buf->target` still comprises a link-local address.

**Comment by BytesGalore** [comment]
@cgundogan yes the commits are ontop of this merge [1] 19 hours ago 

[1] https://github.com/BytesGalore/RIOT/commit/1af61260000844b8c71a629102b4da0b1bb92074

**Comment by cgundogan** [comment]
@BytesGalore could you rebase again please? (: This PR conflicts now with the merged #2607

**Comment by BytesGalore** [comment]
sure, done :)

File: sys/net/routing/rpl/rpl_control_messages.c [hunk 8]
```diff
@@ -386,13 +387,13 @@ void rpl_send_DAO(rpl_dodag_t *my_dodag, ipv6_addr_t *destination, uint8_t lifet
...
-    for (uint8_t i = start_index; i < rpl_max_routing_entries; i++) {
-        if (rpl_get_routing_table()[i].used) {
+    for (uint8_t i = start_index; i < RPL_MAX_PARENTS; i++) {
+        if(rpl_parents[i].used) {
             rpl_send_opt_target_buf->type = RPL_OPT_TARGET;
             rpl_send_opt_target_buf->length = RPL_OPT_TARGET_LEN;
             rpl_send_opt_target_buf->flags = 0x00;
             rpl_send_opt_target_buf->prefix_length = RPL_DODAG_ID_LEN;
-            memcpy(&rpl_send_opt_target_buf->target, &rpl_get_routing_table()[i].address,
+            memcpy(&rpl_send_opt_target_buf->target, &rpl_parents[i].addr,
```
Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: Taking the address from the parent results in these link-local addresses we see. In the former implementation `rpl_get_routing_table()[i].address` returned global addresses. I think the proper way would be to consult the fib and get the global destination address for those link-local addresses, and skip all parents without an entry in the fib.

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: Unfortunately its not possible that the FIB replies a global address for a given next-hop address, since a next-hop address can be easily used for a large number of destinations.
Iterating is also not possible due to the separation concept of the FIB. But even if iterating would be possible, the FIB can have more entries than just the next-hops from RPL parents.

Fortunately not all is lost :D we could keep the global address for a parent in `rpl_parents[i].addr` instead of the local one, and ask the FIB for the global address when needed.

Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: On second thought, iterating over the parents to fill the DAO is not the right way, because DAOs accumulate the routes they encounter up the dodag to the root. By iterating over the parents we can only put their addresses in the DAO but not the addresses the node accumulated through other DAOs (from children). In my opinion there should be a way to iterate over all targets in the fib and write them into the DAO.

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: This way we loose the downward route information for sub-DODAGS right?

Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: I fear so, yes. How could we approach this issue? Is it possible to implement something like a generic iterator for the fib?

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: Technically yes, but I think this is not exactly what we want since the FIB contains probably more destinations then RPL has in use, other DODAGs, entries from other routing protocols, static routes etc.
That tricky.

Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: Just thinking out loud: What about allowing minimal application-specific data in the fib? So that an application could iterate over the set of routes and only pick those that are marked in a specific way by that application?

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: I have to think a bit more on it. 
Just to clarify for myself, what we need to know here are all destinations addresses reachable from this node, which are constructed (storing MOP assumed) by receiving DIOs and all target options from all received DAOs right?

Comment (cgundogan) [inline sys/net/routing/rpl/rpl_control_messages.c]: Correct, a DAO includes the address of the sending node in addition to all addresses it received  from its children's DAOs, recursively. (storing MOP)

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: @cgundogan enhancing the data structures in the FIB or using the type field (`uint32_t`) for additional indication is no problem.
Maybe (and very related to your proposition), and this is also just thinking out loud, the FIB could return a full set of next-hop/destination tuples based on matching type field or just specific bits of it.

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: Maybe something like this:
``` C
typedef struct entries{
    ng_ipv6_addr_t dst;
    ng_ipv6_addr_t next_hop;
};

/**
* param[in]  bitmask indicate the bits to use in the search
* param[in]  typevalue the query type
* param[out] arr_entries array to store matching entries
* param[in, out] arr_entries_size indicates the storage spce on in and the use count when returned
*/
void fib_get_entry_set(uint32_t bitmask, uint32_t typevalue, entries* arr_entries, size_t* arr_entries_size);
```

Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: @cgundogan I've opened #2818 as proposition for a function that could help here

**Comment by gebart** [comment]
needs rebase

**Comment by BytesGalore** [comment]
rebased.

**Comment by BytesGalore** [comment]
Anyone interested to keep this PR?

**Comment by cgundogan** [comment]
@BytesGalore not really (: I will close it as this references the former `rpl` implementation and the new `ng_rpl` implementation has already the `fib` included.
